Automox Plugin: Initial release of plugin with base management functionality

ax-zyoutz opened the pull request:
## Proposed Changes

### Description

Describe the proposed changes:

  - New plugin for management and integration with Automox
  - Included functionality:
    - Device management
    - Group management
    - Policy management
    - Vulnerability Sync

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [x] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [x] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [x] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [x] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [x] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [x] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [x] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [x] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [x] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

ax-zyoutz commented:
@mrinehart-r7 @blaxminarayan-r7 I've pushed the last actions/trigger along with some updates to examples with regeneration. I'm working on the description details but think this is close to done. I've been running the validators and ther are three aciton example to help.md that I can't seem to get past - almost appears to be validator escaping issues.  Any thoughts?

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
Input violations: Action -> "Upload Vulnerability Sync File": Missing ['|csv_file_name|string|insightconnect-uploaded-report.csv|False|Name for CSV file uploaded and shown within Automox|None|insightconnect-uploaded-report.csv|'] in help.md
Input violations: Trigger -> "Get Automox Events": Missing ['|event_type|string|None|True|Name of event type to be retrieved (List of event types found at https://developer.automox.com/openapi/axconsole/operation/getEvents/)|None|user.login|'] in help.md
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpInputOutputValidator" failed! 
        Cause: Help.md is not in sync with plugin.spec.yaml. Please regenerate help.md by running 'icon-plugin generate python --regenerate' to rectify violations.


----
[*] Total time elapsed: 1593.569ms
```

Also, due to the fork/remote and I think not having a previous version I had to comment out the Version Increment Validator

ax-zyoutz commented:
@mrinehart-r7 @blaxminarayan-r7 
Last actions added and updated spec/help/documentation with additional details. Other than the Version validator I was able to get everything else to cleanly pass:
```
zyoutz@zyoutz-MBPro automox % make validate
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 1605.511ms
```

I've also made the updates to test with the slim version of the SDK.

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+products: [insightconnect]
+name: automox
+title: Automox
+description: Automox is Modernizing IT operations with continuous visibility, insight, and agility for your entire IT environment
```
```suggestion
description: Automox is modernizing IT operations with continuous visibility, insight, and agility for your entire IT environment
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+title: Automox
+description: Automox is Modernizing IT operations with continuous visibility, insight, and agility for your entire IT environment
+version: 1.0.0
+supported_versions: ["All"]
```
```suggestion
supported_versions: ["All as of 12/17/2021"]
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+    orgs:
+      title: Organizations
+      description: The organizations for which the user has access
+      type: "[]object"
```
Could this be a []organization? If not, thats fine! Just seeing if we can reuse from above

ax-zyoutz replied:
Unfortunately its different in this context and information can vary

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: The configuration of the policy
+      type: object
+      required: false
+    server_groups:
+      title: Device Groups
+      description: List of device groups assigned policy
+      type: "[]integer"
+      required: false
```
Is this a typo? Field name is server_groups, title/description is device groups. Also please review the data type here- is every device group an integer?

ax-zyoutz replied:
I'm updating the description to be a bit more clear - along with the title. It's a list of Device Group IDs.

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: Enable operating system auto update
+      type: boolean
+      required: false
+    server_count:
+      title: Device Count
```
Again, maybe a bit confusing to have server_count as key here and Device Count as title. Are servers the same as devices?

ax-zyoutz replied:
That is correct. The API in the future will be updated to better reflect but it is accurate that servers/devices are interchangeable. Would you rather have in-action mappings of fields for each of these?

mrinehart-r7 replied:
@ax-zyoutz it should be fine - the user just sees the title properties.

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: Identifier of user
+      type: integer
+      required: false
+    server_id:
+      title: Device ID
```
(see above)

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      required: false
+    policy_name:
+      title: Policy Name
+      description: Name of policydentifier of user
```
```suggestion
      description: Name of policy
```
I think? Unfamiliar with "policydentifier"

mrinehart-r7 replied:
@ax-zyoutz is this a typo?

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: Data included with event
+      type: object
+      required: false
+    server_name:
+      title: Device Name
```
(see above comments)

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      type: integer
+      required: false
+    unknown_host_count:
+      title: Uknown Host Count
```
```suggestion
      title: Unknown Host Count
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      type: string
+      required: false
+    created_by_user:
+      title: Created By User
```
```suggestion
      title: Created by User
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      type: object
+      required: false
+    last_updated_by_user:
+      title: Updated By User
```
```suggestion
      title: Updated by User
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+  api_key:
+    title: API Key
+    type: credential_secret_key
+    description: Enter organization API key
```
```suggestion
    description: Organization API key
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+    input:
+      org_id:
+        title: Organization ID
+        description: Identifier of Organization
```
```suggestion
        description: Identifier of organization
```

pmara-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+        example: 1234
+      group_id:
+        title: Group ID
+        description: Identifier of Server Group
```
```suggestion
        description: Identifier of server group
```

pmara-r7 reviewed (changes requested):
Some initial comments on the plugin spec- also, we believe that you are intentionally leaving some types as object because the API returns varied results, just make sure that is the case as you move forward. If we know the type, define it

jrose-r7 commented on plugins/automox/icon_automox/actions/get_device_by_hostname/action.py:
```diff
@@ -0,0 +1,19 @@
+
+    def run(self, params={}):
+        device = self.connection.automox_api.find_device_by_attribute(params.get(Input.ORG_ID),
+                                                                      ["name"],
```
is the extra "," a typo or apart of convention?

ax-zyoutz replied:
Not sure I follow. There should be three parameters: org id, list of attribute values, and the actual value. Perhaps I'm overlooking the extra ","

jrose-r7 commented on plugins/automox/icon_automox/actions/create_group/action.py:
```diff
@@ -0,0 +1,39 @@
+
+        return {Output.SUCCESS: True, Output.GROUP: self.connection.automox_api.remove_null_values(group)}
+
+    def default_server_group(self, org_id):
```
should this be moved to a util file or connection file and then imported?

jrose-r7 replied:
thanks for clarifying! makes sense

ax-zyoutz commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+enable_cache: false
+
+types:
+  organization:
```
`organization` is the list of type returned for `list_organizations`. Line 414

jrose-r7 replied:
I see! I ctrl + fed for type: organization and missed this. thanks for clarifying!

ax-zyoutz commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: The device name
+      type: string
+      required: true
+  policy:
```
This should be the return of `list_policies` as `"[]policy"`

jrose-r7 replied:
I see! I ctrl + fed for type: policy and missed this. thanks for clarifying!

ax-zyoutz commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -0,0 +1,988 @@
+      description: Number of CVEs that are impacted by batch
+      type: integer
+      required: false
+  task:
```
Returned as list of task for `list_vulnerability_sync_tasks`

mrinehart-r7 commented on plugins/automox/help.md:
```diff
@@ -0,0 +1,1239 @@
+
+# Key Features
+
+* Ability to retrieve and manage Automox managed devices
+* Ability to manage Automox groups
+* Ability to initiate Vulnerability Sync uploads and remediation of issues
+* Ability to trigger workflows based on Automox platform events
```
```suggestion
* Retrieve and manage Automox managed devices
* Manage Automox groups
* Initiate Vulnerability Sync uploads and remediation of issues
* Trigger workflows based on Automox platform events
```

mrinehart-r7 commented on plugins/automox/help.md:
```diff
@@ -0,0 +1,1239 @@
+
+# Requirements
+
+* Requires an Automox API Key
```
```suggestion
* Automox API Key
```

mrinehart-r7 commented on plugins/automox/help.md:
Are there any patch samples you can provide here?

mrinehart-r7 commented on plugins/automox/help.md:
```suggestion
* [Managing API Keys](https://support.automox.com/help/managing-keys)
* [Event Types for Get Automox Events action](https://developer.automox.com/openapi/axconsole/operation/getEvents/#!in=query&path=eventName&t=request)
```

ax-zyoutz commented:
@mrinehart-r7 Don't see your comment but yes `policydentifier` was a typo. Working on a couple types to push right now to avoid a couple more `object` types used. A coupel will still exist since they are dynamic.

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+
+        return page_resp
+
+    def _page_results_data(self, url: str, init_params: Dict = dict()) -> json:
```
```suggestion
    def _page_results_data(self, url: str, init_params: Dict = dict()) -> [dict]:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+            self.logger.info(f"Call to Automox Console API failed: {e}")
+            raise PluginException(preset=PluginException.Preset.UNKNOWN)
+
+    def _page_results(self, url: str, init_params: Dict = dict(), sanitize: bool = True) -> json:
```
```suggestion
    def _page_results(self, url: str, init_params: Dict = dict(), sanitize: bool = True) -> [dict]:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+            "content-type": "application/json"
+        }
+
+    def _call_api(self, method: str, url: str, params: Dict = dict(), json_data: object = None) -> json:
```
```suggestion
    def _call_api(self, method: str, url: str, params: Dict = dict(), json_data: object = None) -> Optional[dict]:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+import io
+import json
+import requests
+from typing import Dict, List
```
```suggestion
from typing import Dict, List, Optional, Collection
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+
+    # Remove Null from response to avoid type issues
+    def remove_null_values(self, d):
+        if type(d) is dict:
+            return dict((k, self.remove_null_values(v)) for k, v in d.items() if v and self.remove_null_values(v))
+        elif type(d) is list:
```
```suggestion
        if isinstance(d, dict):
            return dict((k, self.remove_null_values(v)) for k, v in d.items() if v and self.remove_null_values(v))
        elif isinstance(d, list):
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return page_resp
+
+    # Remove Null from response to avoid type issues
+    def remove_null_values(self, d):
```
```suggestion
    def remove_null_values(self, d: Collection):
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+            return d
+
+    @staticmethod
+    def _org_param(org_id):
```
```suggestion
    def _org_param(org_id: str) -> dict:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return {"o": org_id}
+
+    @staticmethod
+    def first_page(params: Dict = {}):
```
```suggestion
    def first_page(params: Dict = {}) -> None:
```
Depending on if you accept the below change suggestion, you may need to change this return type.

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return {"o": org_id}
+
+    @staticmethod
+    def first_page(params: Dict = {}):
+        params.update({
+            'limit': ApiClient.PAGE_SIZE,
+            'page': 0
+        })
+        return params
```
Dictionaries are pass-by-reference so there's no need to do a return here - you can just do the updates and the original dictionary will be updated.

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+    def get_device(self, org_id: int, device_id: int) -> Dict:
+        return self._call_api("GET", f"{self.endpoint}/servers/{device_id}", params=self._org_param(org_id))
+
+    def find_device_by_attribute(self, org_id: int, attributes: List[str], value: str):
```
```suggestion
    def find_device_by_attribute(self, org_id: int, attributes: List[str], value: str) -> dict:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        params = {"o": org_id, "groupId": group_id}
+        return self._page_results(f"{self.endpoint}/servers", params)
+
+    def run_device_command(self, org_id: int, device_id: int, command: str):
```
```suggestion
    def run_device_command(self, org_id: int, device_id: int, command: str) -> bool:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+
+    # Policies
+    @staticmethod
+    def _sanitize_policies(policies) -> List[Dict]:
```
```suggestion
    def _sanitize_policies(policies: [dict]) -> List[Dict]:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return self._call_api("PUT", f"{self.endpoint}/servergroups/{group_id}",
+                              params=self._org_param(org_id), json_data=payload)
+
+    def delete_group(self, org_id: int, group_id: int):
```
```suggestion
    def delete_group(self, org_id: int, group_id: int) -> bool:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return self._call_api("DELETE", f"{self.endpoint}/servergroups/{group_id}", params=self._org_param(org_id))
+
+    # Vulnerability Sync
+    def upload_vulnerability_sync_file(self, org_id: int, file_content, filename):
```
```suggestion
    def upload_vulnerability_sync_file(self, org_id: int, file_content, filename) -> int:
```

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+                raise PluginException(cause="Failed to upload file to Vulnerability Sync",
+                                      assistance=f"Review encoded CSV file and try again: {e}")
+
+    def get_vulnerability_sync_batches(self, org_id: int):
+        return self._page_results_data(f"{self.endpoint}/orgs/{org_id}/tasks/batches")
+
+    def get_vulnerability_sync_batch(self, org_id: int, batch_id: int):
+        return self._call_api("GET", f"{self.endpoint}/orgs/{org_id}/tasks/batches/{batch_id}")
+
+    def update_vulnerability_sync_batch(self, org_id: int, batch_id: int, action: str):
+        return self._call_api("POST", f"{self.endpoint}/orgs/{org_id}/tasks/batches/{batch_id}/{action}")
+
+    def get_vulnerability_sync_tasks(self, org_id: int, params: Dict):
+        return self._page_results_data(f"{self.endpoint}/orgs/{org_id}/tasks", params)
+
+    def update_vulnerability_sync_task(self, org_id: int, task_id: int, action: str):
+        return self._call_api("PATCH", f"{self.endpoint}/orgs/{org_id}/tasks/{task_id}", params={"action": action})
```
```suggestion
    def get_vulnerability_sync_batches(self, org_id: int) -> bool:
        return self._page_results_data(f"{self.endpoint}/orgs/{org_id}/tasks/batches")

    def get_vulnerability_sync_batch(self, org_id: int, batch_id: int) -> bool:
        return self._call_api("GET", f"{self.endpoint}/orgs/{org_id}/tasks/batches/{batch_id}")

    def update_vulnerability_sync_batch(self, org_id: int, batch_id: int, action: str) -> bool:
        return self._call_api("POST", f"{self.endpoint}/orgs/{org_id}/tasks/batches/{batch_id}/{action}")

    def get_vulnerability_sync_tasks(self, org_id: int, params: Dict) -> bool:
        return self._page_results_data(f"{self.endpoint}/orgs/{org_id}/tasks", params)

    def update_vulnerability_sync_task(self, org_id: int, task_id: int, action: str) -> bool:
        return self._call_api("PATCH", f"{self.endpoint}/orgs/{org_id}/tasks/{task_id}", params={"action": action})
```
Assuming this is bool?

mrinehart-r7 commented on plugins/automox/icon_automox/util/api_client.py:
```diff
@@ -0,0 +1,318 @@
+        return self._call_api("PATCH", f"{self.endpoint}/orgs/{org_id}/tasks/{task_id}", params={"action": action})
+
+    # Events
+    def get_events(self, org_id: int, event_type: str, page: int = 0):
```
```suggestion
    def get_events(self, org_id: int, event_type: str, page: int = 0) -> dict:
```

ax-zyoutz commented:
@mrinehart-r7 @pmara-r7 @jrose-r7  I've gone through feedback and believe I've addressed all feedback.  I also spent some time refining some of the types to ensure there is more useful attributes such as on devices/policies/etc and made sure examples were updated. Actions were tested again to confirm things are working as expected. Couple of things from feedback weren't actual changes needed so I resolved/dismissed those comments.

There was a GTM discussion between Automox/Rapid7 this week and we are ready for this to go out once you all feel comfortable. Once its out I'll get the workflows I have ready updated and get those submitted as well.

mrinehart-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -328,6 +659,48 @@ types:
+  task_payload_package_versions:
+    id:
+      title: Package ID
+      description: The package identifier associated to the task
```
```suggestion
      description: The package identifier associated with the task
```

mrinehart-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -328,6 +659,48 @@ types:
+      required: true
+    name:
+      title: Name
+      description: The name of the package associated to the task
```
```suggestion
      description: The name of the package associated with the task
```

mrinehart-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -328,6 +659,48 @@ types:
+      required: false
+    version:
+      title: Version
+      description: The version of the package associated to the task
```
```suggestion
      description: The version of the package associated with the task
```

mrinehart-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -328,6 +659,48 @@ types:
+      required: false
+    display_name:
+      title: Display Name
+      description: The display name of the package associated to the task
```
```suggestion
      description: The display name of the package associated with the task
```

mrinehart-r7 commented on plugins/automox/plugin.spec.yaml:
```diff
@@ -328,6 +659,48 @@ types:
+      required: false
+    severity:
+      title: Severity
+      description: The severity of the patch associated to the task
```
```suggestion
      description: The severity of the patch associated with the task
```
